File: src/main/kotlin/info/davidvedvick/seis739/catpics/security/CatEmployeeEntry.kt
package info.davidvedvick.seis739.catpics.security

import info.davidvedvick.seis739.catpics.users.CatEmployee
import info.davidvedvick.seis739.catpics.users.ManageCatEmployees

class CatEmployeeEntry(private val manageCatEmployees: ManageCatEmployees, private val passwordEncoder: Encoder) : AuthenticateCatEmployees {

    override suspend fun authenticate(unauthenticatedCatEmployee: UnauthenticatedCatEmployee): CatEmployeeCredentials =
        with (unauthenticatedCatEmployee) {
            val catEmployee = manageCatEmployees.findByEmail(email) ?: manageCatEmployees.save(
                CatEmployee(
                    email = email,
                    password = passwordEncoder.encode(password)
                )
            )

            when {
                !catEmployee.isEnabled -> DisabledCatEmployee(email, password)
                catEmployee.password.isBlank() -> BadCatEmployeeCredentials(email, password)
                passwordEncoder.matches(password, catEmployee.password) -> AuthenticatedCatEmployee(email, password)
                else -> FailedCatEmployeeCredentials(email, password)
            }
        }
}

File: server/security/CatEmployeeEntry.ts
import AuthenticateCatEmployees from "./AuthenticateCatEmployees.js";
import {UnauthenticatedCatEmployee} from "./UnauthenticatedCatEmployee.js";
import {CatEmployeeCredentials} from "./CatEmployeeCredentials.js";
import {ManageCatEmployees} from "../users/ManageCatEmployees.js";
import Encoder from "./Encoder.js";
import DisabledCatEmployee from "./DisabledCatEmployee.js";

export default class CatEmployeeEntry implements AuthenticateCatEmployees {

    constructor(private readonly catEmployees: ManageCatEmployees, private readonly encoder: Encoder) {}

    async authenticate(unauthenticatedCatEmployee: UnauthenticatedCatEmployee): Promise<CatEmployeeCredentials> {
        let catEmployee = await this.catEmployees.findByEmail(unauthenticatedCatEmployee.email);
        if (!catEmployee) {
            catEmployee = await this.catEmployees.save({
                id: 0,
                email: unauthenticatedCatEmployee.email,
                password: this.encoder.encode(unauthenticatedCatEmployee.password),
                isEnabled: false,
            });
        }

        return new DisabledCatEmployee(
            catEmployee.id,
            catEmployee.email,
            catEmployee.password,
            catEmployee.isEnabled,
        );
    }
}